repo2docker/buildpacks/julia/julia_project.py

```diff
@@ -0,0 +1,140 @@
+        if 'compat' in project_toml:
+            if 'julia' in project_toml['compat']:
+                julia_version_str = project_toml['compat']['julia']
+
+                # For Project.toml files, install the latest julia version that
+                # satisfies the given semver.
+                julia_version = find_semver_match(julia_version_str, self.all_julias)
+                if julia_version is None:
+                    return _default_julia_version
+        return default_julia_version
+
+    def get_build_env(self):
+        """Get additional environment settings for Julia and Jupyter
+
+        Returns:
+            an ordered list of environment setting tuples
+
+            The tuples contain a string of the environment variable name and
+            a string of the environment setting:
+            - `JULIA_PATH`: base path where all Julia Binaries and libraries
+                will be installed
+            - `JULIA_HOME`: path where all Julia Binaries will be installed
+            - `JULIA_PKGDIR`: path where all Julia libraries will be installed
+            - `JULIA_DEPOT_PATH`: path where Julia libraries are installed.
+                                  Similar to JULIA_PKGDIR, used in 1.0.
+            - `JULIA_VERSION`: default version of julia to be installed
+            - `JUPYTER`: environment variable required by IJulia to point to
+                the `jupyter` executable
+
+            For example, a tuple may be `('JULIA_VERSION', '0.6.0')`.
+
+        """
+        return super().get_build_env() + [
+            ('JULIA_PATH', '${APP_BASE}/julia'),
+            ('JULIA_HOME', '${JULIA_PATH}/bin'),  # julia <= 0.6
+            ('JULIA_BINDIR', '${JULIA_HOME}'),  # julia >= 0.7
+            ('JULIA_PKGDIR', '${JULIA_PATH}/pkg'),
+            ('JULIA_DEPOT_PATH', '${JULIA_PKGDIR}'), # julia >= 0.7
+            ('JULIA_VERSION', self.julia_version),
+            ('JUPYTER', '${NB_PYTHON_PREFIX}/bin/jupyter')
+        ]
+
+    def get_path(self):
+        """Adds path to Julia binaries to user's PATH.
+
+         Returns:
+             an ordered list of path strings. The path to the Julia
+             executable is added to the list.
+
+        """
+        return super().get_path() + ['${JULIA_HOME}']
+
+    def get_build_scripts(self):
+        """
+        Return series of build-steps common to "ALL" Julia repositories
+
+        All scripts found here should be independent of contents of a
+        particular repository.
+
+        This creates a directory with permissions for installing julia packages
+        (from get_assemble_scripts).
+
+        """
+        return super().get_build_scripts() + [
+            (
+                "root",
+                r"""
+                mkdir -p ${JULIA_PATH} && \
+                curl -sSL "https://julialang-s3.julialang.org/bin/linux/x64/${JULIA_VERSION%[.-]*}/julia-${JULIA_VERSION}-linux-x86_64.tar.gz" | tar -xz -C ${JULIA_PATH} --strip-components 1
```

Comment:
I think this should go into `get_assemble_scripts` because it depends on the contents of the repository.

The build scripts run [before the repo contents is copied over](https://github.com/jupyter/repo2docker/blob/e6cbf3e3a3e399508d5628b4ff1d30afed8c944f/repo2docker/buildpacks/base.py#L91-L96) and [assemble runs after](https://github.com/jupyter/repo2docker/blob/e6cbf3e3a3e399508d5628b4ff1d30afed8c944f/repo2docker/buildpacks/base.py#L131-L135).

We'd like to make sure that all steps before copying over the repository can be cached and shared between repos. So it should be independent of the repo contents.

(I've been pondering this for a while and am starting to think we should put things like installing Julia or RStudio up here. However I'd want to ponder that for a bit longer and get input from others. So for this PR I'd suggest we create a new issue for discussion of this. Otherwise we have to wait on a decision to merge this PR. (There is also the option that my feeling here is off and it is uncontroversial to put this step in the build part, would need some feedback. Maybe @minrk or @yuvipanda have some on how controversial this would be)

Reply:
Looking at the history of `julia/__init__.py` the "install Julia version X" script has been in the build step for a good while, so maybe it is just me who thinks this is controversial. My suggestion is now "Definitely punt any discussion to a new issue".

Reply:
Ok, so we'll leave it as is for now. I agree, I was also not sure what the right strategy is...

Reply:
Sorry that might've been my bad to put it in the wrong spot in the Julia 0.7+ PR...
I also agree that it'd make sense to move it, but I also also agree that it should be done in a new issue. :)